File: frontend/src/services/userService.ts
/**
 * User Service
 */
import { apiClient } from './api';
import type { User, DashboardStats, TypingSession } from '@/types';

interface UserResponse {
  id: string;
  username: string;
  email: string;
  avatar: string | null;
  created_at: string;
  stats: {
    avg_wpm: number;
    avg_accuracy: number;
    best_wpm: number;
    total_sessions: number;
    total_time_typed: number;
    total_characters_typed: number;
  };
}

interface DashboardStatsResponse {
  total_sessions: number;
  avg_wpm: number;
  avg_accuracy: number;
  best_wpm: number;
  total_time_typed: number;
  improvement_percent: number;
  recent_sessions: SessionResponse[];
  wpm_trend: WpmDataPointResponse[];
}

interface SessionResponse {
  id: string;
  text_id: string;
  wpm: number;
  raw_wpm: number;
  accuracy: number;
  errors: number;
  total_characters: number;
  duration: number;
  mode: string;
  started_at: string;
  completed_at: string;
}

interface WpmDataPointResponse {
  time: number;
  wpm: number;
  accuracy: number;
}

function mapUserResponse(response: UserResponse): User {
  return {
    id: response.id,
    username: response.username,
    email: response.email,
    avatar: response.avatar || undefined,
    createdAt: response.created_at,
    stats: {
      avgWpm: response.stats.avg_wpm,
      avgAccuracy: response.stats.avg_accuracy,
      bestWpm: response.stats.best_wpm,
      totalSessions: response.stats.total_sessions,
      totalTimeTyped: response.stats.total_time_typed,
      totalCharactersTyped: response.stats.total_characters_typed,
    },
  };
}

function mapSessionResponse(response: SessionResponse): TypingSession {
  return {
    id: response.id,
    userId: '',
    textId: response.text_id,
    text: '',
    wpm: response.wpm,
    rawWpm: response.raw_wpm,
    accuracy: response.accuracy,
    errors: response.errors,
    totalCharacters: response.total_characters,
    correctCharacters: response.total_characters - response.errors,
    duration: response.duration,
    startedAt: response.started_at,
    completedAt: response.completed_at,
    mode: response.mode as 'solo' | 'private' | 'competition',
    wpmHistory: [],
  };
}

function mapDashboardStats(response: DashboardStatsResponse): DashboardStats {
  return {
    totalSessions: response.total_sessions,
    avgWpm: response.avg_wpm,
    avgAccuracy: response.avg_accuracy,
    bestWpm: response.best_wpm,
    totalTimeTyped: response.total_time_typed,
    improvementPercent: response.improvement_percent,
    recentSessions: response.recent_sessions.map(mapSessionResponse),
    wpmTrend: response.wpm_trend.map((p) => ({
      time: p.time,
      wpm: p.wpm,
      accuracy: p.accuracy,
    })),
  };
}

export const userService = {
  async getCurrentUser(): Promise<User> {
    const response = await apiClient.get<UserResponse>('/users/me');
    return mapUserResponse(response);
  },

  async getDashboardStats(): Promise<DashboardStats> {
    const response = await apiClient.get<DashboardStatsResponse>('/users/me/dashboard');
    return mapDashboardStats(response);
  },

  async updateProfile(data: { username?: string; email?: string; avatar?: string }): Promise<User> {
    const response = await apiClient.put<UserResponse>('/users/me', data);
    return mapUserResponse(response);
  },

  async getUserProfile(userId: string): Promise<User> {
    const response = await apiClient.get<UserResponse>(`/users/${userId}`);
    return mapUserResponse(response);
  },

  async searchUsers(query: string): Promise<Array<{ id: string; username: string; avatar?: string; stats: { avgWpm: number; totalSessions: number } }>> {
    const response = await apiClient.get<Array<{ id: string; username: string; avatar: string | null; stats: { avg_wpm: number; total_sessions: number } }>>(`/users/search?q=${encodeURIComponent(query)}`);
    return response.map((u) => ({
      id: u.id,
      username: u.username,
      avatar: u.avatar || undefined,
      stats: {
        avgWpm: u.stats.avg_wpm,
        totalSessions: u.stats.total_sessions,
      },
    }));
  },
};

File: backend/app/application/services/user_service.py
"""
User service.

Handles user profile management and statistics.
"""
from datetime import UTC, datetime
from uuid import UUID

from app.domain.entities.user import User
from app.domain.exceptions import DuplicateEntityError, EntityNotFoundError
from app.domain.repositories.session_repository import SessionRepository
from app.domain.repositories.user_repository import UserRepository


class UserService:
    """Service handling user operations."""

    def __init__(
        self,
        user_repository: UserRepository,
        session_repository: SessionRepository,
    ):
        self._user_repo = user_repository
        self._session_repo = session_repository

    async def get_user_by_id(self, user_id: UUID) -> User:
        """
        Get a user by their ID.

        Args:
            user_id: User's unique identifier.

        Returns:
            User entity.

        Raises:
            EntityNotFoundError: If user not found.
        """
        user = await self._user_repo.get_by_id(user_id)
        if not user:
            raise EntityNotFoundError("User", str(user_id))
        return user

    async def update_profile(
        self,
        user_id: UUID,
        username: str | None = None,
        email: str | None = None,
        avatar: str | None = None,
    ) -> User:
        """
        Update user profile information.

        Args:
            user_id: User's ID.
            username: New username (optional).
            email: New email (optional).
            avatar: New avatar URL (optional).

        Returns:
            Updated user.

        Raises:
            EntityNotFoundError: If user not found.
            DuplicateEntityError: If new email/username already exists.
        """
        user = await self._user_repo.get_by_id(user_id)
        if not user:
            raise EntityNotFoundError("User", str(user_id))

        if username and username != user.username:
            if await self._user_repo.exists_by_username(username):
                raise DuplicateEntityError("User", "username", username)
            user.username = username.strip()

        if email and email.lower() != user.email:
            email = email.lower().strip()
            if await self._user_repo.exists_by_email(email):
                raise DuplicateEntityError("User", "email", email)
            user.email = email

        if avatar is not None:
            user.avatar = avatar if avatar else None

        user.updated_at = datetime.now(UTC)
        return await self._user_repo.update(user)

    async def delete_account(self, user_id: UUID) -> bool:
        """
        Delete a user account.

        Args:
            user_id: User's ID.

        Returns:
            True if deleted successfully.
        """
        return await self._user_repo.delete(user_id)

    async def search_users(self, query: str, limit: int = 10) -> list[User]:
        """
        Search users by username.

        Args:
            query: Search query.
            limit: Maximum results.

        Returns:
            List of matching users.
        """
        return await self._user_repo.search_by_username(query, limit)

    async def get_user_dashboard_stats(self, user_id: UUID) -> dict:
        """
        Get dashboard statistics for a user.

        Args:
            user_id: User's ID.

        Returns:
            Dictionary with dashboard statistics.
        """
        user = await self.get_user_by_id(user_id)
        stats = await self._session_repo.get_user_stats(user_id)
        recent_sessions = await self._session_repo.get_recent_sessions(user_id, limit=5)

        sessions_week_ago = await self._session_repo.get_user_sessions(
            user_id=user_id,
            limit=100,
            offset=0,
        )

        improvement = 0.0
        if len(sessions_week_ago[0]) >= 2:
            first_half = sessions_week_ago[0][: len(sessions_week_ago[0]) // 2]
            second_half = sessions_week_ago[0][len(sessions_week_ago[0]) // 2 :]

            if first_half and second_half:
                first_avg = sum(s.wpm for s in first_half) / len(first_half)
                second_avg = sum(s.wpm for s in second_half) / len(second_half)
                if second_avg > 0:
                    improvement = ((first_avg - second_avg) / second_avg) * 100

        wpm_trend = []
        for session in reversed(recent_sessions):
            wpm_trend.append(
                {
                    "time": len(wpm_trend) + 1,
                    "wpm": session.wpm,
                    "accuracy": session.accuracy,
                }
            )

        return {
            "total_sessions": stats["total_sessions"],
            "avg_wpm": stats["avg_wpm"],
            "avg_accuracy": stats["avg_accuracy"],
            "best_wpm": user.stats.best_wpm,
            "total_time_typed": stats["total_time_typed"],
            "improvement_percent": round(improvement, 1),
            "recent_sessions": [
                {
                    "id": str(s.id),
                    "wpm": s.wpm,
                    "accuracy": s.accuracy,
                    "duration": s.duration,
                    "mode": s.mode.value,
                    "completed_at": s.completed_at.isoformat(),
                }
                for s in recent_sessions
            ],
            "wpm_trend": wpm_trend,
        }
